File: tutorial/ios/code/ViewController_NASA_GIBS.m
//
//  ViewController.m
//  HelloEarth
//
//  Created by Steve Gifford on 11/11/14.
//  Copyright (c) 2014 mousebird consulting. All rights reserved.
//

#import "ViewController.h"
#import "WhirlyGlobeComponent.h"

@interface ViewController ()
{
    NSMutableDictionary *ovlLayers;
}

- (void) addCountries;

@end

@implementation ViewController
{
    WhirlyGlobeViewController *theViewC;
    NSDictionary *vectorDict;
    CLLocationManager *locationManager;
    NSArray *picsArray;
    NSData *thePicData;
}

// Set these for different view options
const bool DoOverlay = true;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Create an empty globe or map and add it to the view
    theViewC = [[WhirlyGlobeViewController alloc] init];
    [self.view addSubview:theViewC.view];
    theViewC.view.frame = self.view.bounds;
    [self addChildViewController:theViewC];
    
    // this logic makes it work for either globe or map
    WhirlyGlobeViewController *globeViewC = nil;
    MaplyViewController *mapViewC = nil;
    if ([theViewC isKindOfClass:[WhirlyGlobeViewController class]])
        globeViewC = (WhirlyGlobeViewController *)theViewC;
    else
        mapViewC = (MaplyViewController *)theViewC;
    
    // we want a black background for a globe, a white background for a map.
    theViewC.clearColor = (globeViewC != nil) ? [UIColor lightGrayColor] : [UIColor whiteColor];
    
    // and thirty fps if we can get it ­ change this to 3 if you find your app is struggling
    theViewC.frameInterval = 3;     //2
    
// Varies the tilt per height
 //   [theViewC setTiltMinHeight:0.005 maxHeight:0.10 minTilt:1.10 maxTilt:0.02];
    
    // add the capability to use the local tiles or remote tiles
    bool useLocalTiles = false;
    
    // we'll need this layer in a second
    MaplyQuadImageTilesLayer *layer;
    
    if (useLocalTiles)
    {
        MaplyMBTileSource *tileSource =
        [[MaplyMBTileSource alloc] initWithMBTiles:@"geography­-class_medres"];
        layer = [[MaplyQuadImageTilesLayer alloc]
                 initWithCoordSystem:tileSource.coordSys tileSource:tileSource];
    } else {
        // Because this is a remote tile set, we'll want a cache directory
        NSString *baseCacheDir =
        [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES)
         objectAtIndex:0];
        NSString *tilesCacheDir = [NSString stringWithFormat:@"%@/tiles/",
                                         baseCacheDir];

        
// A set of various base layers to select from. Remember to adjust the maxZoom factor appropriately
        // http://tile.stamen.com/terrain/
        // http://map1.vis.earthdata.nasa.gov/wmts-webmerc/VIIRS_CityLights_2012/default/2015-05-07/GoogleMapsCompatible_Level8/{z}/{y}/{x} - jpg
        // http://map1.vis.earthdata.nasa.gov/wmts-webmerc/MODIS_Terra_CorrectedReflectance_TrueColor/default/2015-06-07/GoogleMapsCompatible_Level9/{z}/{y}/{x}  - jpg
        
        int maxZoom = 18;
        
        // Stamen Terrain Tiles, courtesy of Stamen Design under the Creative Commons Attribution License.
        // Data by OpenStreetMap under the Open Data Commons Open Database License.
        MaplyRemoteTileSource *tileSource =
        [[MaplyRemoteTileSource alloc]
         initWithBaseURL:@"http://tile.stamen.com/terrain/"
         ext:@"jpg" minZoom:0 maxZoom:maxZoom];
        tileSource.cacheDir = tilesCacheDir;
        layer = [[MaplyQuadImageTilesLayer alloc]
                 initWithCoordSystem:tileSource.coordSys tileSource:tileSource];
    }
    
    layer.handleEdges = (globeViewC != nil);
    layer.coverPoles = (globeViewC != nil);
    layer.requireElev = false;
    layer.waitLoad = false;
    layer.drawPriority = 0;
    layer.singleLevelLoading = false;
    [theViewC addLayer:layer];
    
    // start up over Santa Cruz, center of the universe's beach
    if (globeViewC != nil)
    {
        globeViewC.height = 0.06;
        globeViewC.heading = 0.15;
        globeViewC.tilt = 0.0;         // PI/2 radians = horizon??
        [globeViewC animateToPosition:MaplyCoordinateMakeWithDegrees(-122.4192,37.7793)
                                 time:1.0];
    } else {
        mapViewC.height = 0.05;
        [mapViewC animateToPosition:MaplyCoordinateMakeWithDegrees(-122.4192,37.7793)
                               time:1.0];
    }
    
    
    // Setup a remote overlay from NASA GIBS
    if (DoOverlay)
    {
        // For network paging layers, where we'll store temp files
        NSString *cacheDir = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES)  objectAtIndex:0];
        
        // Example NASA GIBS overlay layers
        // http://map1.vis.earthdata.nasa.gov/wmts-webmerc/MODIS_Terra_Chlorophyll_A/default/2015-02-10/GoogleMapsCompatible_Level7/{z}/{y}/{x}
        // http://map1.vis.earthdata.nasa.gov/wmts-webmerc/Sea_Surface_Temp_Blended/default/2015-06-25/GoogleMapsCompatible_Level7/{z}/{y}/{x}
        // http://tile.openweathermap.org/map/precipitation/
        
        MaplyRemoteTileSource *tileSource = [[MaplyRemoteTileSource alloc] initWithBaseURL:@"http://map1.vis.earthdata.nasa.gov/wmts-webmerc/Sea_Surface_Temp_Blended/default/2015-06-25/GoogleMapsCompatible_Level7/{z}/{y}/{x}" ext:@"png" minZoom:0 maxZoom:7];
        
        tileSource.cacheDir = [NSString stringWithFormat:@"%@/sea_temperature/",cacheDir];
        
        tileSource.tileInfo.cachedFileLifetime = 3; // invalidate OWM data after three secs
        MaplyQuadImageTilesLayer *temperatureLayer = [[MaplyQuadImageTilesLayer alloc] initWithCoordSystem:tileSource.coordSys tileSource:tileSource];
        
        //       NSLog(@"The coordSystem is %@", tileSource.coordSys);
        
        temperatureLayer.coverPoles = false;
        temperatureLayer.handleEdges = false;
        [globeViewC addLayer:temperatureLayer];
        
    }
    
    // set the vector characteristics to be pretty and selectable
    vectorDict = @{
                   kMaplyColor: [UIColor whiteColor],
                   kMaplySelectable: @(true),
                   kMaplyVecWidth: @(4.0)};
}

- (void)addCountries
{
    // handle this in another thread
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT,0),
                   ^{
                       NSArray *allOutlines = [[NSBundle mainBundle] pathsForResourcesOfType:@"geojson" inDirectory:nil];
                       
                       for (NSString *outlineFile in allOutlines)
                       {
                           NSData *jsonData = [NSData dataWithContentsOfFile:outlineFile];
                           if (jsonData)
                           {
                               MaplyVectorObject *wgVecObj = [MaplyVectorObject VectorObjectFromGeoJSON:jsonData];
                               
                               // the admin tag from the country outline geojson has the country name ­ save
                               NSString *vecName = [[wgVecObj attributes] objectForKey:@"ADMIN"];
                               wgVecObj.userObject = vecName;
                               
                               // add the outline to our view
                               MaplyComponentObject *compObj = [theViewC addVectors:[NSArray arrayWithObject:wgVecObj] desc:vectorDict];
                               // If you ever intend to remove these, keep track of the MaplyComponentObjects above.
             /*                  // Add a screen label per country
                               if ([vecName length] > 0)
                               {
                                   MaplyScreenLabel *label = [[MaplyScreenLabel alloc] init];
                                   label.text = vecName;
                                   label.loc = [wgVecObj center];
                                   label.selectable = true;
                                   [theViewC addScreenLabels:@[label] desc:
                                    @{
                                      kMaplyFont: [UIFont boldSystemFontOfSize:12.0],
                                      kMaplyTextOutlineColor: [UIColor blackColor],
                                      kMaplyTextOutlineSize: @(1.0),
                                      kMaplyColor: [UIColor whiteColor]
                                      }];
              
                               }
                           */
                           }
                       }
                   });
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: tutorial/code/ViewController_NASA_GIBS.swift
//
//  ViewController.swift
//  HelloEarth
//
//  Created by jmnavarro on 24/07/15.
//  Copyright (c) 2015 Mousebird. All rights reserved.
//

import UIKit

class ViewController: UIViewController,
	WhirlyGlobeViewControllerDelegate,
MaplyViewControllerDelegate {

	private var theViewC: WhirlyGlobeViewController?
	private var vectorDict: [String:AnyObject]?

	private let DoOverlay = true


	override func viewDidLoad() {
		super.viewDidLoad()

		theViewC = WhirlyGlobeViewController()
		self.view.addSubview(theViewC!.view)
		theViewC!.view.frame = self.view.bounds
		addChildViewController(theViewC!)

		// we want a black background for a globe, a white background for a map.
		theViewC!.clearColor = UIColor.blackColor()

		// and thirty fps if we can get it ­ change this to 3 if you find your app is struggling
		theViewC!.frameInterval = 3

		// add the capability to use the local tiles or remote tiles
		let useLocalTiles = false

		// we'll need this layer in a second
		let layer: MaplyQuadImageTilesLayer

		if useLocalTiles {
			if let tileSource = MaplyMBTileSource(MBTiles: "geography-class_medres") {
				layer = MaplyQuadImageTilesLayer(tileSource: tileSource)
			}
		}
		else {
			// Because this is a remote tile set, we'll want a cache directory
			let baseCacheDir = NSSearchPathForDirectoriesInDomains(.CachesDirectory, .UserDomainMask, true)[0]
			let aerialTilesCacheDir = "\(baseCacheDir)/osmtiles/"

			// A set of various base layers to select from. Remember to adjust the maxZoom factor appropriately
			// http://otile1.mqcdn.com/tiles/1.0.0/sat/
			// http://map1.vis.earthdata.nasa.gov/wmts-webmerc/VIIRS_CityLights_2012/default/2015-05-07/GoogleMapsCompatible_Level8/{z}/{y}/{x} - jpg
			// http://map1.vis.earthdata.nasa.gov/wmts-webmerc/MODIS_Terra_CorrectedReflectance_TrueColor/default/2015-06-07/GoogleMapsCompatible_Level9/{z}/{y}/{x}  - jpg

			let maxZoom = Int32(18)

			// MapQuest Open Aerial Tiles, Courtesy Of Mapquest
			// Portions Courtesy NASA/JPL­Caltech and U.S. Depart. of Agriculture, Farm Service Agency

			if let tileSource = MaplyRemoteTileSource(
					baseURL: "http://otile1.mqcdn.com/tiles/1.0.0/sat/",
					ext: "jpg",
					minZoom: 0,
					maxZoom: maxZoom) {
				tileSource.cacheDir = aerialTilesCacheDir
				layer = MaplyQuadImageTilesLayer(tileSource: tileSource)
			}
		}


		layer.handleEdges = true
		layer.coverPoles = true
		layer.requireElev = false
		layer.waitLoad = false
		layer.drawPriority = 0
		layer.singleLevelLoading = false
		theViewC!.addLayer(layer)

		// start up over Santa Cruz, center of the universe's beach
		theViewC!.height = 0.06
		theViewC!.heading = 0.15
		theViewC!.tilt = 0.0         // PI/2 radians = horizon??
		theViewC!.animateToPosition(MaplyCoordinateMakeWithDegrees(-122.4192,37.7793), time: 1.0)

		// Setup a remote overlay from NASA GIBS
		if DoOverlay {
			// For network paging layers, where we'll store temp files
			let cacheDir = NSSearchPathForDirectoriesInDomains(.CachesDirectory, .UserDomainMask, true)[0]

			if let tileSource = MaplyRemoteTileSource(baseURL: "http://map1.vis.earthdata.nasa.gov/wmts-webmerc/Sea_Surface_Temp_Blended/default/2013-06-07/GoogleMapsCompatible_Level7/",
					ext: "png",
					minZoom: 1,
					maxZoom: 7) {
				tileSource.cacheDir = "\(cacheDir)/sea_temperature/"
				tileSource.tileInfo.cachedFileLifetime = 3 // invalidate OWM data after 3 secs
				if let temperatureLayer = MaplyQuadImageTilesLayer(tileSource: tileSource) {
					temperatureLayer.coverPoles = false
					temperatureLayer.handleEdges = false
					theViewC!.addLayer(temperatureLayer)
				}
			}
		}

		vectorDict = [
			kMaplyColor: UIColor.whiteColor(),
			kMaplySelectable: true,
			kMaplyVecWidth: 4.0
		]

		// add the countries
		addCountries()
	}


	private func addCountries() {
		// handle this in another thread
		let queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)
		dispatch_async(queue) {
			let allOutlines = NSBundle.mainBundle().pathsForResourcesOfType("geojson", inDirectory: nil)

			for outline in allOutlines {
				if let jsonData = NSData(contentsOfFile: outline),
						wgVecObj = MaplyVectorObject(fromGeoJSON: jsonData) {
					// the admin tag from the country outline geojson has the country name ­ save
					if let attrs = wgVecObj.attributes,
							vecName = attrs.objectForKey("ADMIN") as? NSObject {

						wgVecObj.userObject = vecName

						if count(vecName.description) > 0 {
							let label = MaplyScreenLabel()
							label.text = vecName.description
							label.loc = wgVecObj.center()
							label.selectable = true
							self.theViewC?.addScreenLabels([label],
								desc: [
									kMaplyFont: UIFont.boldSystemFontOfSize(24.0),
									kMaplyTextOutlineColor: UIColor.blackColor(),
									kMaplyTextOutlineSize: 2.0,
									kMaplyColor: UIColor.whiteColor()
								])
						}
					}

					// add the outline to our view
					let compObj = self.theViewC?.addVectors([wgVecObj], desc: self.vectorDict)

					// If you ever intend to remove these, keep track of the MaplyComponentObjects above.
					
				}
			}
		}
	}
	
}
